File: svc-md/src/main/java/com/beancounter/marketdata/service/MdFactory.java
package com.beancounter.marketdata.service;

import com.beancounter.common.model.Asset;
import com.beancounter.common.model.Market;
import com.beancounter.marketdata.providers.alpha.AlphaProviderService;
import com.beancounter.marketdata.providers.mock.MockProviderService;
import com.beancounter.marketdata.providers.wtd.WtdProviderService;
import java.util.HashMap;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

/**
 * Return a MarketData provider from a registered collection.
 *
 * @author mikeh
 * @since 2019-03-01
 */
@Service
@Slf4j
public class MdFactory {

  private Map<String, MarketDataProvider> providers = new HashMap<>();

  @Autowired
  MdFactory(MockProviderService mockProviderService,
            AlphaProviderService alphaProviderService,
            WtdProviderService wtdProviderService) {
    providers.put(mockProviderService.getId().toUpperCase(), mockProviderService);
    providers.put(alphaProviderService.getId().toUpperCase(), alphaProviderService);
    providers.put(wtdProviderService.getId().toUpperCase(), wtdProviderService);
  }

  /**
   * Figures out how to locate a Market Data provider for the requested asset.
   * If one can't be found, then the MOCK provider is returned.
   *
   * @param asset who wants to know?
   * @return the provider that supports the asset
   */
  MarketDataProvider getMarketDataProvider(Asset asset) {
    MarketDataProvider provider = resolveProvider(asset.getMarket());
    if (provider == null) {
      return providers.get(MockProviderService.ID);
    }
    return provider;
  }

  public MarketDataProvider getMarketDataProvider(String provider) {
    return providers.get(provider.toUpperCase());
  }

  private MarketDataProvider resolveProvider(Market market) {
    // ToDo: Map Market to Provider
    if (providers.get(MockProviderService.ID).isMarketSupported(market)) {
      return providers.get(MockProviderService.ID);
    }
    if (providers.get(AlphaProviderService.ID).isMarketSupported(market)) {
      return providers.get(AlphaProviderService.ID);
    }
    if (providers.get(WtdProviderService.ID).isMarketSupported(market)) {
      return providers.get(WtdProviderService.ID);
    }
    log.error("Unable to identify a provider for {}", market);
    return null;
  }


}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/providers/MdFactory.kt
package com.beancounter.marketdata.providers

import com.beancounter.common.model.Market
import com.beancounter.marketdata.providers.alpha.AlphaPriceService
import com.beancounter.marketdata.providers.cash.CashProviderService
import com.beancounter.marketdata.providers.custom.PrivateMarketDataProvider
import com.beancounter.marketdata.providers.marketstack.MarketStackService
import com.beancounter.marketdata.providers.morningstar.MorningstarPriceService
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import java.util.Locale

/**
 * Return a MarketData provider from a registered collection.
 *
 * @author mikeh
 * @since 2019-03-01
 */
@Service
class MdFactory internal constructor(
    cashProviderService: CashProviderService,
    alphaPriceService: AlphaPriceService,
    privateMarketDataProvider: PrivateMarketDataProvider,
    marketStackService: MarketStackService,
    morningstarPriceService: MorningstarPriceService
) {
    private val providers: Map<String, MarketDataPriceProvider> =
        mapOf(
            Pair(
                CashProviderService.ID,
                cashProviderService
            ),
            Pair(
                MarketStackService.ID,
                marketStackService
            ),
            Pair(
                AlphaPriceService.ID,
                alphaPriceService
            ),
            Pair(
                PrivateMarketDataProvider.ID,
                privateMarketDataProvider
            ),
            Pair(
                MorningstarPriceService.ID,
                morningstarPriceService
            )
        )

    /**
     * Locate a price provider for the requested market.
     *
     * @param market who is pricing this for this market
     * @return the provider that supports the asset
     */
    @Cacheable("provider")
    fun getMarketDataProvider(market: Market): MarketDataPriceProvider = resolveProvider(market)

    fun getMarketDataProvider(provider: String): MarketDataPriceProvider =
        providers[provider.uppercase(Locale.getDefault())]!!

    private fun resolveProvider(market: Market): MarketDataPriceProvider {
        // ToDo: Map Market to Provider
        for (key in providers.keys) {
            if (providers[key]!!.isMarketSupported(market)) {
                return providers[key]!!
            }
        }
        log.error(
            "Unable to identify a provider for {}",
            market
        )
        return providers[CashProviderService.ID]!!
    }

    companion object {
        private val log = LoggerFactory.getLogger(MdFactory::class.java)
    }
}